### PR TITLE
Add debugger display for huge BigInteger

### LINKED ISSUE #17453: Optimize BigInteger "DataTip" speed
For very big BigIntegers the Visual Studio DataTip (the thing that appears when hovering over a variable with the mouse) takes a long time to load. I assume that's because the ToString processing takes too long.

The result is not visible anyway because it is aborted by a timeout as it seems.

Can the DataTip be made faster for super long BigIntegers? Maybe it could just show the first and last 20 digits instead of all digits. It's still possible to see all digits by calling ToString using the debugger.

Maybe the output could be `$"{firstDigits} ... {lastDigits} ({totalDigits} digits in total, {byteCount} bytes}"`. That would be useful to both see the data and to get a feel for memory consumption.

The ToString output would stay unchanged.

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Closes #17453
Didn't find a good way to test it. The algorithm is quite simple.
Assuming there won't be leading 0's in `_bits`. Is this expected contract?

### CODE REVIEW

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (kasperk81) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1602]:
```suggestion
            if (_bits is null || _bits.Length <= 15)
```
2 is too small?

**Comment by stephentoub** [comment]
> Didn't find a good way to test it. 

e.g. https://github.com/dotnet/runtime/blob/508fef51e841aa16ffed1aae32bf4793a2cea363/src/libraries/System.Runtime.Extensions/tests/System/Diagnostics/Stopwatch.cs#L103

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (tfenise) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1611]: Are these high bits or low bits?

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (tfenise) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1618]:
```suggestion
            while (significand >= 10.0)
```

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (tfenise) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1615]: Could this overflow for very, very large `BigInteger`s?

**Comment by tfenise** [comment]
Should we make it explicit that the debugger display is only an estimation? It's not floating point where rounding errors are everywhere, and it may be confusing if, e.g., `3e+100` is displayed as `2.9999999e+100`.

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (tfenise) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1627]: Maybe too much of an edge case, but `significand < 10.0` is not enough to guarantee `$"{significand:F8}"` is less than ten.

For example,
```C#
double d = 9.9999999999;
Console.WriteLine(d < 10.0);
Console.WriteLine(d.ToString("G17"));
Console.WriteLine(d.ToString("F8"));
```
outputs
```
True
9.9999999999
10.00000000
```

Comment (tfenise) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1627]: We could instead multiply `significand` by some power of ten, round it to an integer, convert it into a decimal string, count the digits, add the decimal point, remove trailing zeros, ... , to avoid this problem.

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (huoyaoyuan) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1602]: Each element in `_bits` is 32 bits. Starting at 64 bits doesn't look so small.

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (huoyaoyuan) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1615]: Yes it can, beyond 10^(2^32).

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (huoyaoyuan) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1627]: Rounding, dividing by power of ten, and then formatting in double would be sufficient.

Comment (huoyaoyuan) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1627]: Combining with the FP precision problem, I think rounding instead of truncating looks better.

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (tfenise) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1619]:
```suggestion
                long exponent = (long)exponentLow;
```

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (tfenise) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1615]:
```suggestion
                double lowBitsCount32 = _bits.Length - 2; // if Length > int.MaxValue/32, counting in bits can cause overflow
                double exponentLow = lowBitsCount32 * kcbitUint * Math.Log10(2);
```
otherwise `lowBitsCount32 * kcbitUint` is calculated with `int`.

Maybe worth adding an analyzer someday. `int * int * double` may not do what one really wants.

**Review by tfenise — CHANGES_REQUESTED** [review]

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (huoyaoyuan) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1615]: `lowBitsCount32` can also be `long`, but using `double` is fine either.

**Comment by huoyaoyuan** [comment]
> System.InvalidOperationException : Expected one DebuggerDisplayAttribute on System.Numerics.BigInteger.

Why is the debugger display attribute removed in browser? Is it fine to skip the test on browser then?

**Comment by stephentoub** [comment]
> Why is the debugger display attribute removed in browser?

They get stripped away by the linker due to https://github.com/dotnet/runtime/blob/e117479e2530ed4e0e83e3e27948b5675fda2b57/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.Shared.xml#L10-L12

>  Is it fine to skip the test on browser then?

Yes

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (dakersnar) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1614]: Is this overflow possible, given we limit the length of the BigInteger's `_bits` array to `Array.MaxLength/sizeof(uint32)`? https://source.dot.net/#System.Runtime.Numerics/System/Numerics/BigInteger.cs,624

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (dakersnar) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1621]: See above, I believe the max possible size of BigInteger is smaller. @tannergooding can you confirm?

Regardless, even if I'm correct and you ease up on these restrictions, please use debug asserts to assert these assumptions.

**Review by dakersnar — COMMENTED** [review]
Looks good, thanks for the contribution! I added a couple points to address. This should also get a second review from @tannergooding.

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (huoyaoyuan) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1614]: I didn't see the limit. If so, overflow won't be possible.

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (danmoseley) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1600]:
```suggestion
                // Display a fast estimated value instead
```

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (danmoseley) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1629]:
```suggestion
                // Keep some digits in the significand. 8 is arbitrarily chosen, about half of the precision of double
```

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (tfenise) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1642]: Probably need to use invariant culture to pass the tests.
```suggestion
                return string.Create(CultureInfo.InvariantCulture, $"{signStr}{significand:F8}e+{exponent}");
```

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (tfenise) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1620]: No. `_bits.Length <= Array.MaxLength / sizeof(uint)` only guarantees the amount of _bytes_ is no more than `Array.MaxLength`, not the amount of _bits_.
```suggestion
                double lowBitsCount32 = _bits.Length - 2; // max length of _bits is Array.MaxLength / sizeof(uint), which unfortunately may have more than int.MaxValue bits.
                double exponentLow = lowBitsCount32 * kcbitUint * Math.Log10(2);

                // Max possible value around 256^Array.MaxLength > 10^int.MaxValue
                // int may overflow.
                long exponent = (long)exponentLow;
                double significand = (double)highBits * Math.Pow(10, exponentLow - exponent);
```

**Review by tfenise — CHANGES_REQUESTED** [review]

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (dakersnar) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs]: Ahh, good point. `sizeof(uint)` is in units of bytes, so it returns `4`, not `32`.

Comment (tannergooding) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs]: I think the issue is that the logic around `MaxLength` matches the initial fix discussed (which is that any `BigInteger` should be able to fit into a `Span<byte>`) but not the secondary fix that was discussed in that we shouldn't have more than `int.MaxValue` bits so that the generic math `GetShortestBitLength` API could do an altogether simpler thing.

As it is currently, the maximum length is `0x1FFF_FFF1` uints, which comes up to `0x7FFF_FFC4` bytes (an approx. `2 GB` allocation) or `0x3_FFFF_FE20` bits.

The requirement that we fit into `int.MaxValue` bits should put the maximum length at `0x3FF_FFFF` uints, which comes up to `0xFFF_FFFC` bytes (an approx. `256 GB` allocation) or `0x7FFF_FFE0` bits.

A `256 GB` allocation for `BigInteger` should be enough for anyone. This is an astronomically big number at `(2^2_147_483_616) - 1` or `~4.10 * 10^646_456_983` and is really pushing the boundaries of what can fit in memory for anything outside supercomputers. Additionally, once you are creating GC allocations this large, you have many other concerns to consider and a different approach using `native memory` is probably better ;)

Comment (dakersnar) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs]: It sounds like during the BigInteger refactor we should come to a consensus on the actual max size (maybe `int.MaxValue` bits) and have every API assume and enforce that cap.

**Comment by huoyaoyuan** [comment]
Any issue remaining in this PR? The failure looks unrelated.

**Review by dakersnar — APPROVED** [review]
Approving, but needs a second review. Thanks for the contribution!

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (tannergooding) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1604]: Nit: add parens:
```suggestion
                if ((_bits is null) || (_bits.Length <= 2))
```

Why `2` rather than say `4` since we have `Int128`/`UInt128` and efficient display for those?

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (tannergooding) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1611]: nit: Add spaces to help with readability
```suggestion
                // Estimate the value x as `L * 2^n`, while L is the value of high bits, and n is the length of low bits
                // Represent L as `k * 10^i`, then `x = L * 2^n = k * 10^(i + (n * log10(2)))`
                // Let `m = n * log10(2)`, the final result would be `x = (k * 10^(m - [m])) * 10^(i+[m])`
```

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (tannergooding) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1615]: A helper constant would make this slightly more efficient. We use `const double Log10V2 = 0.3010299956639812` elsewhere

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (tannergooding) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1627]: Why not `exponent += Floor(Log10(significand))`?

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (tannergooding) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1643]: It might be desirable to make it clear this is "approximate".

Comment (tannergooding) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1643]: Also why string.Create and not just returning the interpolated string to match what `ToString()` does?

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (huoyaoyuan) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1604]: Just arbitrarily selected. I didn't think about this.

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (huoyaoyuan) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1643]: It's should be more natural to be culture-aware, but how to correctly test it?

Comment (tannergooding) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1646]: I believe we have a culture aware CI run. CC. @jeffhandley ?

Comment (tarekgh) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1646]: I would recommend having tests runs with some specific cultures that can have different behaviors.

Comment (huoyaoyuan) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1646]: I think I should take a look about how other culture-aware `ToString` are tested.

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs
Comment (dakersnar) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1630]: What if the significand is less than 1 before line 1626? It won't end up in that range, right?

Comment (huoyaoyuan) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:1630]: Is there a contract that `highBits` should be at least 2^32? (e.g., no leading 0s in LE representation)

**Review by dakersnar — COMMENTED** [review]
Looks good. Added a couple suggestions per @tannergooding's review.

Also, he mentioned that "It might be desirable to make it clear this is "approximate".", and that conversation should be resolved before we merge this. FWIW, I second that suggestion, but I'm not sure how we would do so.

**Comment by dakersnar** [comment]
@huoyaoyuan friendly reminder that this PR has some unresolved suggestions.

**Review by dakersnar — APPROVED** [review]
Approved, LGTM. Could use one more approval from @tannergooding.

**Comment by jeffhandley** [comment]
There's a test failing, @huoyaoyuan. After that's fixed, we can go ahead and merge this, as @tannergooding might not be able to review it soon but they've previously reviewed and provided feedback that was addressed.

❌ System.Numerics.Tests.DebuggerDisplayTests.TestDebuggerDisplay(bits: [0, 0, 1], displayString: \"1.84467441e+19\")
```
Exception Message
Assert.Equal() Failure
           ↓ (pos 1)
Expected: 1.84467441e+19
Actual:   18446744073709551616
           ↑ (pos 1)
```

**Comment by huoyaoyuan** [comment]
Yes, the test case needs change after updating the formatting threshold.

File: src/libraries/System.Runtime.Numerics/tests/BigInteger/DebuggerDisplayTests.cs
Comment (huoyaoyuan) [inline src/libraries/System.Runtime.Numerics/tests/BigInteger/DebuggerDisplayTests.cs:16]:
```suggestion
        [InlineData(new uint[] { 0, 0, 1 }, "18446744073709551616")]
        [InlineData(new uint[] { 0, 0, 0, 0, 1 }, "3.40282367e+38")]
        [InlineData(new uint[] { 0, 0x12345678, 0, 0xCC00CC00, 0x80808080 }, "3.15085891e+57")]
```

**Review by dakersnar — APPROVED** [review]

File: src/libraries/System.Runtime.Numerics/tests/BigInteger/DebuggerDisplayTests.cs
Comment (huoyaoyuan) [inline src/libraries/System.Runtime.Numerics/tests/BigInteger/DebuggerDisplayTests.cs:16]:
```suggestion
        [InlineData(new uint[] { 0, 0x12345678, 0, 0xCC00CC00, 0x80808080 }, "7.33616508e+47")]
```

**Comment by huoyaoyuan** [comment]
All the tests should have passed now.

**Pull request merged**
